File: statistics/fmsketch.go
// Copyright 2017 PingCAP, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package statistics

import (
	"hash"
	"sync"

	"github.com/pingcap/errors"
	"github.com/pingcap/tidb/sessionctx/stmtctx"
	"github.com/pingcap/tidb/types"
	"github.com/pingcap/tidb/util/codec"
	"github.com/pingcap/tipb/go-tipb"
	"github.com/twmb/murmur3"
	"golang.org/x/exp/maps"
)

var murmur3Pool = sync.Pool{
	New: func() any {
		return murmur3.New64()
	},
}

var fmSketchPool = sync.Pool{
	New: func() any {
		return &FMSketch{
			hashset: make(map[uint64]bool),
			maxSize: 0,
		}
	},
}

// FMSketch is used to count the number of distinct elements in a set.
type FMSketch struct {
	hashset map[uint64]bool
	mask    uint64
	maxSize int
}

// NewFMSketch returns a new FM sketch.
func NewFMSketch(maxSize int) *FMSketch {
	result := fmSketchPool.Get().(*FMSketch)
	result.maxSize = maxSize
	return result
}

// Copy makes a copy for current FMSketch.
func (s *FMSketch) Copy() *FMSketch {
	if s == nil {
		return nil
	}
	result := NewFMSketch(s.maxSize)
	for key, value := range s.hashset {
		result.hashset[key] = value
	}
	result.mask = s.mask
	return result
}

// NDV returns the ndv of the sketch.
func (s *FMSketch) NDV() int64 {
	if s == nil {
		return 0
	}
	return int64(s.mask+1) * int64(len(s.hashset))
}

func (s *FMSketch) insertHashValue(hashVal uint64) {
	if (hashVal & s.mask) != 0 {
		return
	}
	s.hashset[hashVal] = true
	if len(s.hashset) > s.maxSize {
		s.mask = s.mask*2 + 1
		for key := range s.hashset {
			if (key & s.mask) != 0 {
				delete(s.hashset, key)
			}
		}
	}
}

// InsertValue inserts a value into the FM sketch.
func (s *FMSketch) InsertValue(sc *stmtctx.StatementContext, value types.Datum) error {
	bytes, err := codec.EncodeValue(sc, nil, value)
	if err != nil {
		return errors.Trace(err)
	}
	hashFunc := murmur3Pool.Get().(hash.Hash64)
	hashFunc.Reset()
	defer murmur3Pool.Put(hashFunc)
	_, err = hashFunc.Write(bytes)
	if err != nil {
		return errors.Trace(err)
	}
	s.insertHashValue(hashFunc.Sum64())
	return nil
}

// InsertRowValue inserts multi-column values to the sketch.
func (s *FMSketch) InsertRowValue(sc *stmtctx.StatementContext, values []types.Datum) error {
	b := make([]byte, 0, 8)
	hashFunc := murmur3Pool.Get().(hash.Hash64)
	hashFunc.Reset()
	defer murmur3Pool.Put(hashFunc)
	for _, v := range values {
		b = b[:0]
		b, err := codec.EncodeValue(sc, b, v)
		if err != nil {
			return err
		}
		_, err = hashFunc.Write(b)
		if err != nil {
			return err
		}
	}
	s.insertHashValue(hashFunc.Sum64())
	return nil
}

// MergeFMSketch merges two FM Sketch.
func (s *FMSketch) MergeFMSketch(rs *FMSketch) {
	if s == nil || rs == nil {
		return
	}
	if s.mask < rs.mask {
		s.mask = rs.mask
		for key := range s.hashset {
			if (key & s.mask) != 0 {
				delete(s.hashset, key)
			}
		}
	}
	for key := range rs.hashset {
		s.insertHashValue(key)
	}
}

// FMSketchToProto converts FMSketch to its protobuf representation.
func FMSketchToProto(s *FMSketch) *tipb.FMSketch {
	protoSketch := new(tipb.FMSketch)
	if s != nil {
		protoSketch.Mask = s.mask
		for val := range s.hashset {
			protoSketch.Hashset = append(protoSketch.Hashset, val)
		}
	}
	return protoSketch
}

// FMSketchFromProto converts FMSketch from its protobuf representation.
func FMSketchFromProto(protoSketch *tipb.FMSketch) *FMSketch {
	if protoSketch == nil {
		return nil
	}
	sketch := fmSketchPool.Get().(*FMSketch)
	sketch.mask = protoSketch.Mask
	for _, val := range protoSketch.Hashset {
		sketch.hashset[val] = true
	}
	return sketch
}

// EncodeFMSketch encodes the given FMSketch to byte slice.
func EncodeFMSketch(c *FMSketch) ([]byte, error) {
	if c == nil {
		return nil, nil
	}
	p := FMSketchToProto(c)
	protoData, err := p.Marshal()
	return protoData, err
}

// DecodeFMSketch decode a FMSketch from the given byte slice.
func DecodeFMSketch(data []byte) (*FMSketch, error) {
	if data == nil {
		return nil, nil
	}
	p := &tipb.FMSketch{}
	err := p.Unmarshal(data)
	if err != nil {
		return nil, errors.Trace(err)
	}
	fm := FMSketchFromProto(p)
	fm.maxSize = 10000 // TODO: add this attribute to PB and persist it instead of using a fixed number(executor.maxSketchSize)
	return fm, nil
}

// MemoryUsage returns the total memory usage of a FMSketch.
func (s *FMSketch) MemoryUsage() (sum int64) {
	// In FMSketch, we will ignore the memory usage of `hashFunc`.
	// As for the variables mask(uint64) and maxSize(int) each will consume 8 bytes. This is the origin of the constant 16.
	// And for the variables hashset(map[uint64]bool), each element in map will consume 9 bytes(8[uint64] + 1[bool]).
	sum = int64(16 + 9*len(s.hashset))
	return
}

func (s *FMSketch) reset() {
	maps.Clear(s.hashset)
	s.mask = 0
	s.maxSize = 0
}

// DestroyAndPutToPool resets the FMSketch and puts it to the pool.
func (s *FMSketch) DestroyAndPutToPool() {
	if s == nil {
		return
	}
	s.reset()
	fmSketchPool.Put(s)
}

File: rust/crates/tidb-stats/src/fmsketch.rs
// Copyright 2026 PingCAP, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! Raw hash-set Flajolet-Martin sketch geometry from
//! `pkg/statistics/fmsketch.go`.
//!
//! The Go owner also hashes `types.Datum` values through tablecodec, feeds
//! sketches from the row sampler, and persists them through tipb protobufs.
//! This leaf starts at an already-owned `u64` hash and therefore keeps those
//! datatype, sampler, protobuf, and statistics-handle seams explicit.

use std::collections::HashSet;

/// Go's fixed fallback used by `DecodeFMSketch` after protobuf restoration.
pub const MAX_SKETCH_SIZE: usize = 10_000;

/// The dependency-closed core of TiDB's FM sketch.
///
/// The sketch keeps unique hashes that satisfy the current mask.  When the
/// set grows past `max_size`, the mask advances from `m` to `2*m + 1` and
/// hashes that do not have all newly-required zero bits are removed.  This is
/// the exact level transition in Go, including wrapping `u64` arithmetic.
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct FmSketch {
    hashes: HashSet<u64>,
    mask: u64,
    max_size: usize,
}

impl FmSketch {
    /// Creates an empty sketch with the source hash-set threshold.
    #[must_use]
    pub fn new(max_size: usize) -> Self {
        Self {
            hashes: HashSet::with_capacity(max_size),
            mask: 0,
            max_size,
        }
    }

    /// Returns the current mask.  A value of `2^r - 1` means that hashes need
    /// at least `r` trailing zero bits to remain in the set.
    #[must_use]
    pub const fn mask(&self) -> u64 {
        self.mask
    }

    /// Returns the source hash-set threshold.
    #[must_use]
    pub const fn max_size(&self) -> usize {
        self.max_size
    }

    /// Returns the number of retained unique hashes.
    #[must_use]
    pub fn len(&self) -> usize {
        self.hashes.len()
    }

    /// Returns whether no hash is retained.
    #[must_use]
    pub fn is_empty(&self) -> bool {
        self.hashes.is_empty()
    }

    /// Returns a stable view of retained hashes for persistence owners.
    ///
    /// The source map is intentionally unordered; callers that serialize the
    /// set must choose and document their own deterministic ordering.
    #[must_use]
    pub fn contains(&self, hash: u64) -> bool {
        self.hashes.contains(&hash)
    }

    /// Returns the source's estimated distinct-value count.
    #[must_use]
    pub fn ndv(&self) -> i64 {
        (self.mask.wrapping_add(1) as i64).wrapping_mul(self.hashes.len() as i64)
    }

    /// Returns the source's portable sketch memory estimate.
    #[must_use]
    pub fn memory_usage(&self) -> u64 {
        16_u64.wrapping_add(8_u64.wrapping_mul(self.hashes.len() as u64))
    }

    /// Inserts an already-hashed value using the source level transition.
    pub fn insert_hash(&mut self, hash: u64) {
        // Go's `(hashVal & mask) != 0` is the complete admission rule.
        if hash & self.mask != 0 {
            return;
        }

        self.hashes.insert(hash);
        if self.hashes.len() > self.max_size {
            self.mask = self.mask.wrapping_mul(2).wrapping_add(1);
            let mask = self.mask;
            self.hashes.retain(|candidate| candidate & mask == 0);
        }
    }

    /// Inserts several already-hashed values in source stream order.
    pub fn insert_hashes<I>(&mut self, hashes: I)
    where
        I: IntoIterator<Item = u64>,
    {
        for hash in hashes {
            self.insert_hash(hash);
        }
    }

    /// Merges another sketch into this one.
    ///
    /// Go first raises the destination mask to the source mask and filters
    /// existing values, then replays source hashes through normal insertion.
    /// There is no dimension check: different `max_size` values are valid and
    /// the destination's threshold controls subsequent transitions.
    pub fn merge(&mut self, source: &Self) {
        if self.mask < source.mask {
            self.mask = source.mask;
            let mask = self.mask;
            self.hashes.retain(|candidate| candidate & mask == 0);
        }
        for &hash in &source.hashes {
            self.insert_hash(hash);
        }
    }
}
